### PR TITLE
Add Header Maps to Build it with Bazel

### PR DESCRIPTION
**Pull request by naik-sanju**
- Headermaps are required to build with Bazel, Incase of XBS Xcode creates this internally for us.
- Headermap is a map data structure which contains mapping of header files to its path . 

Ex: {'MyHeader.h' : '//path-to-MyHeader.h'}

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,6 +11,11 @@ FLEX (Flipboard Explorer) is a set of in-app debugging and exploration tools for
 <img alt="Demo" width=36% height=36% src=https://user-images.githubusercontent.com/8371943/70185687-e842c800-16af-11ea-8ef9-9e071380a462.gif>
 
 
+## How to generate headermaps?
+- Headermaps are required to build with Bazel
+- Since this repository is not in active development so we are manually checking-in generated header maps. If this repo goes in active development We can setup lanes to generate header maps.
+- At present, headermap files are generated using this rule - https://github.com/bazel-ios/rules_ios/blob/f5fd9581d154647bf92ef76ff65dd5175e7c88f4/rules/hmap.bzl#L97
```
Comment (asifmohd) [inline README.md]: Can we add 1-2 lines of code to invoke to generate header map file manually in this README.md?

File: README.md [hunk 2]
```diff
@@ -11,6 +11,11 @@ FLEX (Flipboard Explorer) is a set of in-app debugging and exploration tools for
 <img alt="Demo" width=36% height=36% src=https://user-images.githubusercontent.com/8371943/70185687-e842c800-16af-11ea-8ef9-9e071380a462.gif>
 
 
+## How to generate headermaps?
+- Headermaps are required to build with Bazel
+- Since this repository is not in active development so we are manually checking-in generated header maps. If this repo goes in active development We can setup lanes to generate header maps.
```
Comment (asifmohd) [inline README.md]: >  Since this repository is not in active development so we are manually checking-in generated header maps. If this repo goes in active development We can setup lanes to generate header maps.

I don't think it's fair to call that this repo is "not in active development". As the upstream authors still ship patches and xcode update fixes.

We can just say something like: "we haven't automated this step for now. Here are the steps to do this manually (Patches to automate this setup are welcome)"

File: FLEX.podspec [hunk 3]
```diff
@@ -40,6 +40,7 @@ Pod::Spec.new do |spec|
   spec.xcconfig = {
        'CLANG_CXX_LANGUAGE_STANDARD' => 'gnu++11',
   }
+  spec.preserve_paths = '**/*.hmap',
```
Comment (asifmohd) [inline FLEX.podspec]: Can we refer to the explicit headermaps here?

Looks like there are only two header map files.

Or if we want to use a wildcard, can we restrict it to `*.hmap` so that it only looks for them in the root directory and not subdirectories?

File: .gitignore [hunk 4]
```diff
@@ -13,7 +13,6 @@ xcuserdata
 *.xccheckout
 *.moved-aside
 DerivedData
-*.hmap
```
Comment (asifmohd) [inline .gitignore:16]: I wonder why did the upstream authors add it here.

Is there a chance this hmap file can be overwritten during some kind of xcode build? 🤔 

According to your description in the README.txt it looks like this needs to be generated manually using the bazel rule.

Do you know what would be the Xcode equivalent way of generating this file?

**Review by asifmohd — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (naik-sanju) [reply README.md]: I tried generating using the above rule and then even copying artifacts generated from rules_ios, and  taking it from rules_ios seems good, because while generating on our own we need to do diff and calculate private headers on our own. So In the future to automate this is what I was thinking.

- Create an Example App and add this Pod.
- Build it with Bazel using rules_ios 
- Then copy generated generated hmap files and place it them in root.

Should I update these same steps?

File: FLEX.podspec (hunk 3 above)
Comment (naik-sanju) [reply FLEX.podspec]: Ok I m fine with mentioning it explicit also, but even this `*.hmap` should do? So tomorrow for any reason we add more hmap's or change names we don't have to modify.

And I think *.hmap won't have that much performance downside?

File: .gitignore (hunk 4 above)
Comment (naik-sanju) [reply .gitignore:16]: Xcode generates this `.hmap` and mostly might keeps it in the DerivedData folder I think. It may not keep it in the project sources directory.

Let me buid the example app in this repo and check if it overwrites this file

File: README.md (hunk 2 above)
Comment (naik-sanju) [reply README.md]: I meant since we haven't updated in a long time, but yeah I get it - can be updated in the future. Let me update the wording accordingly.

File: FLEX.podspec (hunk 3 above)
Comment (asifmohd) [reply FLEX.podspec]: > So tomorrow for any reason we add more hmap's or change names we don't have to modify.

I feel it's a good practice to be explicit here. So that when we do start shipping new/updated hmap files, (say due to changes from upstream), then we should check if the changes are still compatible with Bazel or we need to change some rules somewhere.

But I do see your point, I guess if we start shipping more `hmap` files as part of the cocoapod, there shouldn't be any harm as such.

So I think just `*.hmap` works 👍

File: README.md (hunk 1 above)
Comment (asifmohd) [reply README.md]: I am mostly looking for what the full bazel/bazelisk command one needs to invoke to generate this command.

Let's briefly mention that the sample app needs to have a valid WORKSPACE file with the `rules_ios` rule.

And then devs need to execute this CLI command to get the hmap:

Something like:

`bazelisk build FLEX//:FLEX --generate-hmap`

File: .gitignore (hunk 4 above)
Comment (naik-sanju) [reply .gitignore:16]: Confirmed, Xcode keeps them in DerivedData and the .hmaps are not overwritten
Screenshot 2023-09-04 at 12 57 40 PM
Screenshot 2023-09-04 at 12 58 19 PM

File: README.md (hunk 1 above)
Comment (naik-sanju) [reply README.md]: Done. Setup WORKSPACE and BUILD files. also added bazel build command to run in readMe.

- Only copying generated files is a manual step now/

File: README.md (hunk 2 above)
Comment (naik-sanju) [reply README.md]: Done.

File: FLEX.podspec (hunk 3 above)
Comment (naik-sanju) [reply FLEX.podspec]: Done.

File: Example/FLEXample/SupportingFiles/Info.plist [hunk 5]
```diff
@@ -29,23 +29,6 @@
 	<string>NSPhotoLibraryAddUsageDescription</string>
 	<key>NSPhotoLibraryUsageDescription</key>
 	<string>Save images</string>
-	<key>UIApplicationSceneManifest</key>
```
Comment (naik-sanju) [inline Example/FLEXample/SupportingFiles/Info.plist]: This works only with `SceneDelegate` configured, else it throws error hence deleted SceneDelegate related plist properties.

File: FLEX.xcodeproj/project.xcworkspace/contents.xcworkspacedata [hunk 6]
```diff
@@ -1,7 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<Workspace
```
Comment (asifmohd) [inline FLEX.xcodeproj/project.xcworkspace/contents.xcworkspacedata]: Any reason to delete this file?

File: Example/FLEXample/SupportingFiles/FLEXample-Bridging-Header.h
Comment (asifmohd) [inline Example/FLEXample/SupportingFiles/FLEXample-Bridging-Header.h:1]: In the README.md we've mentioned that the project doesn't build right now using Bazel, but it generates the headermap.

So are these file changes renames required to be merged?

If we revert these changes, it'd be easier to keep the fork up to date later on.

And we can mention it in the README.md, that one should rename this folder to remove space before proceeding with the build.  wdyt?

File: Example/FLEXample/SupportingFiles/Info.plist (hunk 5 above)
Comment (asifmohd) [reply Example/FLEXample/SupportingFiles/Info.plist]: Let's include this as a major change in the MR description and why we had to do so.

Possible to add the error message as well?

File: README.md [hunk 7]
```diff
@@ -11,6 +11,15 @@ FLEX (Flipboard Explorer) is a set of in-app debugging and exploration tools for
 <img alt="Demo" width=36% height=36% src=https://user-images.githubusercontent.com/8371943/70185687-e842c800-16af-11ea-8ef9-9e071380a462.gif>
 
 
+## How to generate headermaps?
+- Headermaps are required to build with Bazel
+- To generate a new set of headermaps, cd into Example folder
+    - Run `pod install`
+    - Comment out `xcconfig = {"CLANG_CXX_LANGUAGE_STANDARD": "gnu++11"},` this line from Pods/FLEX/BUILD.bazel file
+    - Run this command `bazel build --config=Debug --ios_multi_cpus=sim_arm64 --xcode_version=14.3.0 --verbose_failures FLEXExample` , right now the build is failing but it generates headermaps.
```
Comment (asifmohd) [inline README.md:19]: This looks good, thanks for adding the WORKSPACE file and everything 🙇 👍 ❤️

File: Example/Podfile.lock [hunk 8]
```diff
@@ -0,0 +1,22 @@
+PODS:
+  - FLEX (4.7.0)
```
Comment (asifmohd) [inline Example/Podfile.lock]: Is there any particular need to commit this file?

Let's keep this in gitignore if possible.

This way, we don't have to worry about updating this file once your changes are merged into `remove_private_api_usage`

File: Example/Podfile [hunk 9]
```diff
@@ -1,6 +1,14 @@
...
+    'apple_framework' => { load: '@build_bazel_rules_ios//rules:framework.bzl', rule: 'apple_framework' }.freeze,
+    'ios_application' => { load: '@build_bazel_rules_ios//rules:app.bzl', rule: 'ios_application' }.freeze,
+    'ios_unit_test' => { load: '@build_bazel_rules_ios//rules:test.bzl', rule: 'ios_unit_test' }.freeze
+  }.freeze,
+}
+
 target 'FLEXample' do
   project 'FLEXample-Cocoapods'
-  pod 'FLEX', :path => '../'
+  pod 'FLEX', :git => 'https://github.com/gojekfarm/Flex.git', :branch => 'add-header-maps', :modular_headers => true , :configurations => ['Debug']
```
Comment (asifmohd) [inline Example/Podfile:13]: To ensure backward compatibility and no changes to existing users.

Let's merge this change into a new branch called `remove-private-api-usage-and-add-header-maps`

Because at the moment, anyone who runs pod repo update or pod install after deleting podfile.lock will get these changes included, which may cause issues for some of other apps.

Let me know what you think. If you feel this change should be harmless, we can also merge it into the existing `remove_private_api_usage` branch.

**Review by asifmohd — CHANGES_REQUESTED** [review]

File: FLEX.xcodeproj/project.xcworkspace/contents.xcworkspacedata (hunk 6 above)
Comment (naik-sanju) [reply FLEX.xcodeproj/project.xcworkspace/contents.xcworkspacedata]: Reverted.

File: Example/FLEXample/SupportingFiles/Info.plist (hunk 5 above)
Comment (naik-sanju) [reply Example/FLEXample/SupportingFiles/Info.plist]: It was failing because of having this variable `PRODUCT_MODULE_NAME` in Info.plist file, Bazel doesn't support this variable by default hence replaced it with `PRODUCT_NAME`. And we don't have to delete this entire key:value pair, reverted the deletion. 

More info on why bazel deosn't support `PRODUCT_MODULE_NAME` by default - https://github.com/bazelbuild/rules_apple/blob/master/doc/common_info.md#variable-substitution

File: Example/FLEXample/SupportingFiles/FLEXample-Bridging-Header.h
Comment (naik-sanju) [reply Example/FLEXample/SupportingFiles/FLEXample-Bridging-Header.h:1]: Done, reverted space change, So spaces in folders are problem only for `bazel coverage` command, bazel build works fine with spaces also, hence reverted the changes.

File: Example/Podfile (hunk 9 above)
Comment (naik-sanju) [reply Example/Podfile:13]: Yeah let's keep this in a separate branch that way it would be easy to revert also if something goes wrong. Ok keeping these changes in `remove-private-api-usage-and-add-header-maps`

File: Example/Podfile.lock (hunk 8 above)
Comment (naik-sanju) [reply Example/Podfile.lock]: Done. reverted

**Comment by naik-sanju** [comment]
@asifmohd Pushed changes to new branch here - https://github.com/gojekfarm/FLEX/compare/remove_private_api_usage...gojekfarm:FLEX:remove-private-api-usage-and-add-header-maps?expand=1 , once you review changes in this MR. This MR can be closed.

**Comment by asifmohd** [comment]
One more ask @naik-sanju 🙏 

Can we squash all these commits in the new branch?

It would make it easy for us to rebase in future, when we include latest changes from upstream.

File: Example/FLEXample/Supporting Files/Info.plist [hunk 10]
```diff
@@ -41,7 +41,7 @@
 					<key>UISceneConfigurationName</key>
 					<string>Default Configuration</string>
 					<key>UISceneDelegateClassName</key>
-					<string>$(PRODUCT_MODULE_NAME).SceneDelegate</string>
+					<string>$(PRODUCT_NAME).SceneDelegate</string>
```
Comment (asifmohd) [inline Example/FLEXample/Supporting Files/Info.plist:44]: Can you leave a comment in this thread as to why this change was required?

File: Example/Podfile (hunk 9 above)
Comment (asifmohd) [inline Example/Podfile:13]: let's change this to point to the new branch as well

File: Example/FLEXample/Supporting Files/Info.plist (hunk 10 above)
Comment (naik-sanju) [reply Example/FLEXample/Supporting Files/Info.plist:44]: Bazel doesn't support this variable by default because, for targets like AppExtensions, We create 2 targets in Bazel and that creates 2 module names.
- `ios_extension` 
- and the underlying `swift_library`, so it can not figure out which module name to use hence they recommend explicitly setting value for this in Plist instead of using this variable , more info - https://github.com/bazelbuild/rules_apple/blob/master/doc/common_info.md#variable-substitution

Comment (naik-sanju) [reply Example/FLEXample/Supporting Files/Info.plist:44]: Since most of the cases PRODUCT_MODULE_NAME & PRODUCT_NAME would be same unless we explicitly override module name, hence I am using PRODUCT_NAME variable here.
